### PR TITLE
Mertens Function

### PR DESCRIPTION
**Pull request by malayaleecoder**
Added Mertens Function, see https://en.wikipedia.org/wiki/Mertens_function for more details.
Please review and suggest me further improvements!

### CODE REVIEW

**Comment by Sumith1896** [comment]
Do add tests too.

File: symengine/ntheory.cpp [hunk 1]
```diff
@@ -1498,4 +1498,15 @@ int mobius(const Integer &a)
     }
 }
 
+int mertens(const Integer &a){
+    if(a.as_int() <= 0){
+        throw std::runtime_error("mertens: Integer <= 0");
```
Comment (Sumith1896) [inline symengine/ntheory.cpp]: Improve the message

File: symengine/ntheory.cpp [hunk 2]
```diff
@@ -1498,4 +1498,15 @@ int mobius(const Integer &a)
...
+    }
+    unsigned mertens = 0;
+    for(unsigned i = 1; i <= a.as_int(); i++){
+        RCP<const Integer> j = integer(i);
+        mertens = mertens + mobius(*j);
```
Comment (certik) [inline symengine/ntheory.cpp]: You can write these two lines as:
```
mertens += mobius(*integer(i));
```

File: symengine/ntheory.cpp [hunk 3]
```diff
@@ -1498,4 +1498,15 @@ int mobius(const Integer &a)
     }
 }
 
+int mertens(const Integer &a){
+    if(a.as_int() <= 0){
```
Comment (certik) [inline symengine/ntheory.cpp]: Add space after `if` and before `{`.

File: symengine/ntheory.cpp [hunk 4]
```diff
@@ -1498,4 +1498,15 @@ int mobius(const Integer &a)
...
+        throw std::runtime_error("mertens: Integer <= 0");
+    }
+    unsigned mertens = 0;
+    for(unsigned i = 1; i <= a.as_int(); i++){
```
Comment (certik) [inline symengine/ntheory.cpp]: Add space after `if` and before `{`.

**Comment by isuruf** [comment]
Is this a widely used function? I don't see it in SymPy nor Sage

**Comment by malayaleecoder** [comment]
@Sumith1896 @certik: Any more improvisations required? And why is Appveyor and Travis CI still showing that the build is in progress?

**Comment by malayaleecoder** [comment]
@isuruf : I don't think it is widely used. I heard about this function from it's related conjecture, https://en.wikipedia.org/wiki/Mertens_conjecture.

File: symengine/ntheory.cpp [hunk 5]
```diff
@@ -1498,4 +1498,15 @@ int mobius(const Integer &a)
     }
 }
 
+int mertens(const Integer &a)
+{
+    if(a.as_int() <= 0) {
```
Comment (certik) [inline symengine/ntheory.cpp]: Space after `if`.

File: symengine/ntheory.cpp [hunk 6]
```diff
@@ -1498,4 +1498,15 @@ int mobius(const Integer &a)
...
+        throw std::runtime_error("mertens : Mertens function is only defined for positive integers!");
+    }
+    unsigned mertens = 0;
+    for(unsigned i = 1; i <= a.as_int(); i++) {
```
Comment (certik) [inline symengine/ntheory.cpp]: space after `for`.

**Comment by certik** [comment]
Given the way the function is implemented, it only works if the argument fits into an int. Why not to then change the argument from `const Integer &a` to `const long int a`.

And change the tests to just `REQUIRE(mertens(31) == -4);`.

It's not widely used, but I think there is no harm in having it, as long as it is well tested. Also can you add a test for the negative integer? Test that it raises an exception.

**Comment by malayaleecoder** [comment]
@certik : Do verify and let me know about any improvements :)

**Comment by Sumith1896** [comment]
Looks good. You can squash your commits.

**Comment by malayaleecoder** [comment]
@Sumith1896 : Done. Please check!

File: symengine/ntheory.cpp [hunk 7]
```diff
@@ -1498,4 +1498,15 @@ int mobius(const Integer &a)
     }
 }
 
+int mertens(const int long a)
```
Comment (isuruf) [inline symengine/ntheory.cpp]: This can just be `unsigned long` and then you don't have to check for negative values

**Comment by malayaleecoder** [comment]
@isuruf : If I make it unsigned long, should I add tests where negative values throw an exception?

**Comment by certik** [comment]
If you make it unsigned, I don't think you can assign a negative value to it, can you? So that seems like a nice solution.

**Comment by isuruf** [comment]
+1

**Comment by certik** [comment]
This looks very simple and nice. +1 from me as well.

File: symengine/ntheory.cpp [hunk 8]
```diff
@@ -1498,4 +1498,12 @@ int mobius(const Integer &a)
     }
 }
 
+int mertens(const unsigned long a)
+{
+    unsigned mertens = 0;
+    for (unsigned i = 1; i <= a; i++) {
```
Comment (isuruf) [inline symengine/ntheory.cpp]: Either all of `a, mertens, i` should be `unsigned` or `unsigned long`

Comment (certik) [inline symengine/ntheory.cpp]: I missed that, yes.

Comment (malayaleecoder) [inline symengine/ntheory.cpp]: Sorry for missing that.

**Comment by malayaleecoder** [comment]
@isuruf @certik : Should I squash the commits again and make it final?

File: symengine/ntheory.cpp [hunk 9]
```diff
@@ -1498,4 +1498,12 @@ int mobius(const Integer &a)
     }
 }
 
+int mertens(const unsigned long a)
+{
+    unsigned long mertens = 0;
```
Comment (Sumith1896) [inline symengine/ntheory.cpp]: summation of `mobius` can be negative. `unsigned long` is not the right type.

Comment (srajangarg) [inline symengine/ntheory.cpp]: Your function returns `int` but inside, you have returned an `unsigned long`, the test cases passes only because the answer is `int` casted.

**Comment by malayaleecoder** [comment]
I think this should fix it. Please have a look.

**Comment by malayaleecoder** [comment]
@certik : Is there any more improvements to be implemented? Please check.

**Comment by certik** [comment]
@malayaleecoder it looks good to me. Would you mind squashing the commits? The final change is very simple, I think a single commit is fine.

**Comment by malayaleecoder** [comment]
@certik I am unable to squash the commits in this branch. So I am closing this and am creating a new PR.

**Comment by certik** [comment]
That works. Thanks!
